airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams/async_job_manager.py

```diff
@@ -0,0 +1,167 @@
+
+from .async_job import AsyncJob, ParentAsyncJob
+
+if TYPE_CHECKING:
+    from source_facebook_marketing.api import API
+
+logger = logging.getLogger("airbyte")
+
+
+class InsightAsyncJobManager:
+    """
+    Class for managing Ads Insights async jobs. Before running next job it
+    checks current insight throttle value and if it greater than THROTTLE_LIMIT
+    variable, no new jobs added.
+    To consume completed jobs use completed_job generator, jobs will be returned in the same order they finished.
+    """
+
+    # When current insights throttle hit this value no new jobs added.
+    THROTTLE_LIMIT = 70
+    MAX_NUMBER_OF_ATTEMPTS = 5
+    # Time to wait before checking job status update again.
+    JOB_STATUS_UPDATE_SLEEP_SECONDS = 30
+    # Maximum of concurrent jobs that could be scheduled. Since throttling
+    # limit is not reliable indicator of async workload capability we still
+    # have to use this parameter. It is equal to maximum number of request in batch (FB API limit)
+    MAX_JOBS_IN_QUEUE = 100
+    MAX_JOBS_TO_CHECK = 50
+
+    def __init__(self, api: "API", jobs: Iterator[AsyncJob]):
+        """Init
+
+        :param api:
+        :param jobs:
+        """
+        self._api = api
+        self._jobs = jobs
+        self._running_jobs = []
+
+    def _start_jobs(self):
+        """Enqueue new jobs."""
+
+        self._update_api_throttle_limit()
+        self._wait_throttle_limit_down()
+        prev_jobs_count = len(self._running_jobs)
+        while self._get_current_throttle_value() < self.THROTTLE_LIMIT and len(self._running_jobs) < self.MAX_JOBS_IN_QUEUE:
+            job = next(iter(self._jobs), None)
+            if not job:
+                self._empty = True
+                break
+            job.start()
+            self._running_jobs.append(job)
+
+        logger.info(
+            f"Added: {len(self._running_jobs) - prev_jobs_count} jobs. "
+            f"Current throttle limit is {self._current_throttle()}, "
+            f"{len(self._running_jobs)}/{self.MAX_JOBS_IN_QUEUE} job(s) in queue"
+        )
+
+    def completed_jobs(self) -> Iterator[AsyncJob]:
+        """Wait until job is ready and return it. If job
+            failed try to restart it for FAILED_JOBS_RESTART_COUNT times. After job
+            is completed new jobs added according to current throttling limit.
+
+        :yield: completed jobs
+        """
+        if not self._running_jobs:
+            self._start_jobs()
+
+        while self._running_jobs:
+            completed_jobs = self._check_jobs_status_and_restart()
+            while not completed_jobs:
+                logger.info(f"No jobs ready to be consumed, wait for {self.JOB_STATUS_UPDATE_SLEEP_SECONDS} seconds")
+                time.sleep(self.JOB_STATUS_UPDATE_SLEEP_SECONDS)
+                completed_jobs = self._check_jobs_status_and_restart()
+            yield from completed_jobs
+            self._start_jobs()
```

Comment:
wouldn't this have been called above before? Why are we calling it again? would this cause duplicate output?

Reply:
the first call happens before the loop, to ensure we started any job in the beginning.
the second call happens after we potentially had some jobs finished.
`completed_jobs` is a generator so the first call will happen only once.

Reply:
I see. So this is like `kickoff_ready_jobs` rather than a start ALL jobs. SGTM